### PR TITLE
Alasdair/ecom 2913 revised app architecture

### PR DESCRIPTION
**Pull request by AlasdairSwan**
@rlucioni @jimabramson please review this PR.

### CODE REVIEW

File: programs/static/js/routers/dev_app_program_router.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+        'js/routers/core_app_program_router',
+        'js/views/program_list_view'
+    ],
+    function( ProgramsListModel, AdminAppRouter, ProgramListView ) {
+        'use strict';
+
+        return AdminAppRouter.extend({
+            routes: {
+                '': 'programList',
+                'home': 'programList',
```
Comment (rlucioni) [inline programs/static/js/routers/dev_app_program_router.js]: Is defining a route for "home" necessary?

Comment (AlasdairSwan) [inline programs/static/js/routers/dev_app_program_router.js]: Maybe it isn't actually. For local dev I can just set the data attribute to be /author

File: programs/static/js/routers/core_app_program_router.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        'js/views/program_list_view',
+        'js/models/pagination_model',
+        'js/models/program_model'
+    ],
+    function( Backbone, ProgramCreatorView, ProgramDetailsView,
+              ProgramListView, ProgramsListModel, ProgramModel ) {
+        'use strict';
+
+        return Backbone.Router.extend({
+            root: '/author/',
```
Comment (rlucioni) [inline programs/static/js/routers/core_app_program_router.js]: The Studio index page is served at "/home," courses are edited at "/course," and libraries are edited at "/library." Given that, it seems like the root for this app should be "/program." The dev app can override this if necessary, or we can rejigger the Django routes.

Comment (AlasdairSwan) [inline programs/static/js/routers/core_app_program_router.js]: I'm fine with the dev app having the same url structure so will rejigger the Django routes.

File: programs/static/js/views/program_admin_app_view_core.js [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        function( Backbone, ProgramRouter ) {
+            return Backbone.View.extend({
+                el: '.js-program-admin',
+
+                events: {
+                    'click .js-app-click': 'navigate'
+                },
+
+                initialize: function() {
+                    var home = this.$el.data('home-url');
```
Comment (rlucioni) [inline programs/static/js/views/program_admin_app_view_core.js]: Would it be appropriate to include logic for processing a second data attribute containing a URL for the Programs API in this PR? Right now, it's assumed to be `/api/v1/programs/`, but we'll want to be able to specify a fully-qualified URL (e.g., `https://programs.edx.org/api/v1/programs/`) via a data attribute.

Comment (AlasdairSwan) [inline programs/static/js/views/program_admin_app_view_core.js]: I'm thinking it may be better to keep that as a separate PR / Jira story for the sake of cleaness.

File: programs/static/js/routers/core_app_program_router.js (hunk 2 above)
Comment (rlucioni) [inline programs/static/js/routers/core_app_program_router.js]: Cool, let me know if you need a hand. Changing the root to `/program` also means the routes below can be shortened to `/new` and `/:id`.

Comment (AlasdairSwan) [inline programs/static/js/routers/core_app_program_router.js]: Yep, thanks.

File: programs/static/js/views/program_admin_app_view_core.js (hunk 3 above)
Comment (rlucioni) [inline programs/static/js/views/program_admin_app_view_core.js]: Okay, I just don't want to forget about it. It seems like part of this rearchitecture, so maybe it'd be appropriate as a subtask of ECOM-2913.

Comment (AlasdairSwan) [inline programs/static/js/views/program_admin_app_view_core.js]: That works.

File: programs/static/js/apps/program_admin_app.js [hunk 4]
```diff
@@ -1,71 +1,14 @@
...
 
+    /**
+     * Allows for rapid toggling between app version locally
+     * i.e. just swap out the file name in the require array
+     */
     require([
-            'backbone',
-            'js/models/pagination_model',
-            'js/routers/program_router'
+            'js/views/program_admin_app_view_dev'
```
Comment (jimabramson) [inline programs/static/js/apps/program_admin_app.js]: is this the right default value for this?

File: programs/static/js/apps/program_admin_app_core.js [hunk 5]
```diff
@@ -0,0 +1,9 @@
+define([
+        'js/views/program_admin_app_view_core'
+    ],
+    function( ProgramAdminApp ) {
+        'use strict';
+
+        return new ProgramAdminApp();
+    }
+);
```
Comment (jimabramson) [inline programs/static/js/apps/program_admin_app_core.js]: any reason app_core and app_dev have different structures?

File: programs/static/js/routers/dev_app_program_router.js [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    ],
+    function( ProgramsListModel, AdminAppRouter, ProgramListView ) {
+        'use strict';
+
+        return AdminAppRouter.extend({
+            routes: {
+                '': 'programList',
+                'new': 'programCreator',
+                ':id': 'programDetails'
+            },
```
Comment (jimabramson) [inline programs/static/js/routers/dev_app_program_router.js]: very minor, but is it possible / sane to append the programList to the parent's routes instead of duplicating?

**Comment by jimabramson** [comment]
@AlasdairSwan code looks good AFAICT, but the commit message needs to be fixed, and a request would be for some brief documentation (for future maintainers) that explains why things have been factored this way.

File: programs/urls.py [hunk 7]
```diff
@@ -60,7 +60,7 @@
 
     urlpatterns += [
         # Drops into the Programs authoring app, which handles its own routing.
-        url(r'^author/*', author_view, name='author'),
+        url(r'^program/*', author_view, name='author'),
```
Comment (jimabramson) [inline programs/urls.py]: maybe we should rename this from author to admin_dev or something similar, to be consistent with the underlying renames?

File: programs/static/js/apps/program_admin_app_core.js (hunk 5 above)
Comment (AlasdairSwan) [inline programs/static/js/apps/program_admin_app_core.js]: Because mistakes were made (by me) :)

Comment (AlasdairSwan) [inline programs/static/js/apps/program_admin_app_core.js]: Mistakes which have since been rectified.

**Comment by AlasdairSwan** [comment]
@jimabramson re. your comment around brief documentation. Where would said documentation live? I am assuming it is explaining the core vs dev app approach.

**Comment by rlucioni** [comment]
@AlasdairSwan I'd put that documentation in programs/docs, so that it ends up [here](https://open-edx-programs-guide.readthedocs.org/en/latest/).

**Comment by jimabramson** [comment]
@AlasdairSwan yes that is the doc topic I meant.  I would have said to put inline doc, but agree with @rlucioni's better idea to use programs/docs, since the thing you'll document is really spread across many files.

File: programs/static/js/routers/dev_app_program_router.js (hunk 6 above)
Comment (AlasdairSwan) [inline programs/static/js/routers/dev_app_program_router.js]: It would be perfectly sane if Backbone would allow such an obvious thing easily (which I have just discovered it does not). Hopefully I can come back to this later and clean up if I find a solution, but I don't want to spin my wheels too long for such a minor update.

**Comment by AlasdairSwan** [comment]
@jimabramson @rlucioni I have made the relevant changes, can you please give it one more look over and let me know if it's good to go?

File: docs/programs.rst [hunk 8]
```diff
@@ -2,3 +2,9 @@ Programs
 ========
 
 Programs are collections of courses codes offered together. Each "course code" in a Program consists of a collection of course runs, specific instances of a course which a student may take on their way to completing a Program. Together with rules governing which courses may be added to a Program, Programs can be used to support product lines like edX's XSeries offering.
+
+Local Development
+-----------------
+
+You may notice that there are 2 versions of the app (core & dev).
+The core version includes the ability to create new programs and view existing onces. The dev version extends this to also include a programs list view.
```
Comment (rlucioni) [inline docs/programs.rst]: s/onces/ones.

File: docs/programs.rst [hunk 9]
```diff
@@ -2,3 +2,9 @@ Programs
 ========
 
 Programs are collections of courses codes offered together. Each "course code" in a Program consists of a collection of course runs, specific instances of a course which a student may take on their way to completing a Program. Together with rules governing which courses may be added to a Program, Programs can be used to support product lines like edX's XSeries offering.
+
+Local Development
+-----------------
+
+You may notice that there are 2 versions of the app (core & dev).
```
Comment (rlucioni) [inline docs/programs.rst]: As someone from the open source coming to this with no prior experience, I would appreciate knowing what "the app" is. I'd rename this subsection "Authoring App" and add a brief explanation of what the app is for before explaining the differences between the two versions.

File: programs/static/js/views/program_admin_app_view_core.js (hunk 3 above)
Comment (rlucioni) [inline programs/static/js/views/program_admin_app_view_core.js]: I've created https://openedx.atlassian.net/browse/ECOM-3034 to track this.

**Comment by rlucioni** [comment]
Looking good @AlasdairSwan, I just made a few minor comments about documentation. Rebasing onto master once https://github.com/edx/programs/pull/54 lands should get your build passing again.

**Comment by AlasdairSwan** [comment]
@rlucioni now with better documentation, can you please take another look?

File: docs/programs.rst [hunk 10]
```diff
@@ -2,3 +2,20 @@ Programs
 ========
 
 Programs are collections of courses codes offered together. Each "course code" in a Program consists of a collection of course runs, specific instances of a course which a student may take on their way to completing a Program. Together with rules governing which courses may be added to a Program, Programs can be used to support product lines like edX's XSeries offering.
+
+Authoring App
+-----------------
+This application is for creating and managing groupings of courses, which we refer to as **programs**.
+
+
+You may notice that there are 2 versions of the application (core & dev). The core version includes the ability to create new programs and view existing onces. The dev version extends this to also include a programs list view. So the contents of each app are;
```
Comment (rlucioni) [inline docs/programs.rst]: Minor, but it behooves me to point out that this semicolon should really be a colon.

**Comment by rlucioni** [comment]
:+1:

File: docs/programs.rst [hunk 11]
```diff
@@ -2,3 +2,20 @@ Programs
 ========
 
 Programs are collections of courses codes offered together. Each "course code" in a Program consists of a collection of course runs, specific instances of a course which a student may take on their way to completing a Program. Together with rules governing which courses may be added to a Program, Programs can be used to support product lines like edX's XSeries offering.
+
+Authoring App
+-----------------
+This application is for creating and managing groupings of courses, which we refer to as **programs**.
+
+
+You may notice that there are 2 versions of the application (core & dev). The core version includes the ability to create new programs and view existing onces. The dev version extends this to also include a programs list view. So the contents of each app are:
```
Comment (jimabramson) [inline docs/programs.rst]: s/onces/ones

**Comment by jimabramson** [comment]
:+1:  but is that the intended commit message?

**Comment by AlasdairSwan** [comment]
Yes, I like to list the main parts of the PR to help me remember what I did.
